Polish Testing Smart Contracts page (#970)
casper-network/docs: source/docs/casper/developers/writing-onchain-code/testing-contracts.md

- `casper-execution-engine` - This crate imports the execution engine functionality, enabling Wasm execution within the test framework. Each node contains an instance of an execution engine, and the testing framework simulates this behavior.
- `casper-engine-test-support` - A helper crate that provides the interface to write tests and interact with an instance of the execution engine.
- `casper-types` - Types shared by many Casper crates for use on a Casper network.

> ## Setting up the Debugger and types of debugging

Reviewer: I suggest moving this entire section (lines 35 to 108) into a new document and placing it under Resources. Also include some additional links to Rust resources and VSCode documentation.
This document, `testing-contracts`, focuses only on the Casper setup needed to test a Casper contract, and assumes developers landing on this page already know how to debug. This content is useful though, but I see it fitting more under Resources.

Reviewer: Also, please run the content through a peer editor or a writing assistant like Grammarly next time.

There are many ways to make sure that the code is written properly and will not cause any problems after being deployed on the blockchain. While having a good understating of Casper testing framework is crucial for a proper smart contract development, it is also necessary to know how to leverage other tools which will confirm that the logic in the smart contract is functioning properly.
It was already mentioned that debugging in Rust is not active out of the box. There are some steps that need to be completed in order to be able to debug code in Rust.

### Debugging code using breakpoints in VS Code:

> 1. Set up CodeLLDB extension to be able to set up breakpoints in the code. This is a native debugger supporting C++, Rust and other compiled languages.
> 2. Go into the main project folder and type the command “make test” in the VS Code Terminal. This will create the target folder.
> 3. Mark the tests.rs and from the toolbar in VS Code go to Run -> Start Debugging. This will create a new configuration for Rust.
> 4. You will be able to set breakpoints now.
> 5. There are two possibilities to run unit tests:
> - Either from the main project folder using the command “make test”
> - Or from the “tests” subfolder using the command “cargo test”

Reviewer: ```suggestion
1. Set up the CodeLLDB extension, a native debugger supporting C++, Rust, and other compiled languages. This way, you can debug your code by adding breakpoints.
2. Select the main project folder. Type the `make test` command in the VS Code terminal to create the target folder with Wasm files.
3. Select the tests.rs file. Using the toolbar, create a new configuration for Rust by navigating to `Run -> Start Debugging`.
4. You will be able to set breakpoints now.
5. There are two possibilities to run unit tests:
- Either from the main project folder using the command `make test`
- Or from the `tests` subfolder using the command `cargo test`
```

> Both variants will yield the same result which will run the tests and show if all were completed successfully.

Reviewer: ```suggestion
Both variants will yield the same result, running the tests and showing if all were completed successfully.
```

<p align="center"><img src={useBaseUrl("/image/testing-contracts/running-tests.png")} alt="running-tests" width="600"/></p>

> If you are completely sure that the code written in the unit tests is correct, then there is no further need to go deeper than that. It is advisable however to know what to do if you want to know how the code exactly works.

Reviewer: This explanation needs to be clarified. What exactly do you mean here? Can you reword it?

The debug configuration needs to be set up in the following way, so it allows for the debugging of the tests.

```bash
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "lldb",
            "request": "launch",
            "name": "Debug",
            "program":  "${workspaceRoot}/default-project/tests/target/debug/deps/integration_tests-2771dd9bbf2fac3c",
            "args": [],
            "cwd": "${workspaceRoot}",
            "sourceLanguages": [
                "rust"
            ]
        }
    ]
}
```

> The most important is the program path which needs to be set exactly to the unit tests path seen during the “make test” or “cargo test” unit tests run.

Reviewer: Note that we don't use double quotes to highlight, but single quotes. Also, try to avoid passive voice whenever you can.
```suggestion
You need to set the program path to match the path for running the tests; for example, use the path where you ran the tests using `make test` or `cargo test`.
```

> In this case this would be `integration_tests-2771dd9bbf2fac3c`.

Reviewer: ```suggestion
In this example, the path is set to: "${workspaceRoot}/default-project/tests/target/debug/deps/integration_tests-2771dd9bbf2fac3c".
```

6.	Set the breakpoints in the unit tests file.

<p align="center"><img src={useBaseUrl("/image/testing-contracts/set-breakpoint.png")} alt="set-breakpoint" width="600"/></p>

> 7.	Select the test.rs file and run the program in debug mode (F5).

Reviewer: ```suggestion
7.	Select the `test.rs` file and run the program in debug mode (key shortcut: F5).
```

<p align="center"><img src={useBaseUrl("/image/testing-contracts/caught-breakpoint.png")} alt="caught-breakpoint" width="600"/></p>

> This will allow you to see the parameters of the debugged unit tests.

Reviewer: ```suggestion
This will allow you to see the unit test parameters while debugging.
```

### Debugging with macros

> A second most popular type of debugging is putting the “println!” Macro in the code so the parameters can be printed out during “cargo test”.

Reviewer: ```suggestion
The second most popular type of debugging is using the `println!` macro in the code so the parameters can be printed out while running `cargo test`.
```

<p align="center"><img src={useBaseUrl("/image/testing-contracts/println-macro.png")} alt="println-macro" width="600"/></p>

> It is important to note, that running “cargo test” without any parameters will not log the “println!” on the screen.

Reviewer: ```suggestion
It is important to note that running `cargo test` without any parameters will not log the `println!` in the terminal.
```

For this to take effect the command

```bash
cargo test -- --nocapture
```

> must be used.

Reviewer: For this to take effect, the following command must be used:

```bash
cargo test -- --nocapture
```

This will produce the following outcome:

<p align="center"><img src={useBaseUrl("/image/testing-contracts/println-tests.png")} alt="println-tests" width="600"/></p>

> One thing that might be confusing is that although the macro is put in the first test function, it appears on top of the test stack. This is due to Rust running the tests in parallel.

Reviewer: ```suggestion
It might be confusing that although the macro is put in the first test function, it appears on top of the test stack. This is due to Rust running the tests in parallel.
```

> Those kinds of tests should be run targeting the function in question as to not introduce too much confusion during analysis.

Reviewer: ```suggestion
Those tests should target the function in question to make sure everything is clear during analysis.
```
